addons/telemetry-coverage-bug1487578/bootstrap.js

```diff
@@ -0,0 +1,89 @@
+/* -*- indent-tabs-mode: nil; js-indent-level: 2 -*- */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+let {classes: Cc, interfaces: Ci, utils: Cu} = Components;
+
+/* eslint-disable mozilla/use-chromeutils-import */
+Cu.import("resource://gre/modules/AppConstants.jsm");
+Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/ClientID.jsm");
+
+Cu.importGlobalProperties(["crypto", "TextEncoder", "FormData", "fetch"]);
+
+// we want to control this on the client rather than
+// depending on server-side throttling, as throttling
+// cannot account for any other concurrent gradual roll-outs.
+const ENABLE_PROB = 0.01;
+const DEBUG = false;
+const OPT_OUT_PREF = "toolkit.telemetry.coverage.opt-out";
+const TELEMETRY_ENABLED_PREF = "toolkit.telemetry.enabled";
+const REPORTING_ENDPOINT = "https://telemetry-coverage.mozilla.org/submit/coverage/coverage/1";
+
+function debug(msg, obj) {
+  if (DEBUG) {
+    console.log(`Telemetry measurement: ${msg}`); // eslint-disable-line no-console
+    if (obj) {
+        console.log("Telemetry measurement JS object:", obj);
+    }
+  }
+}
+
+async function reportTelemetrySetting() {
+  // this is a locked pref so it *really should* be present, but on the off chance it is not let's assume false.
+  let enabled = Services.prefs.getBoolPref(TELEMETRY_ENABLED_PREF, false);
+
+  const payload = {
+    "appVersion": Services.appinfo.version,
+    "appUpdateChannel": Services.appinfo.defaultUpdateChannel,
```

Comment:
This does not necessarily match what we report in Telemetry. Our starting point is [here](https://searchfox.org/mozilla-central/rev/5a18fb5aeeec99f1ca1c36a697082c221189a3b9/toolkit/components/telemetry/TelemetryController.jsm#320).

Not matching this might be fine, but should be called out for analysis as it might make complicate direct comparisons.

Reply:
That is helpful, thanks!

Reply:
Hm, so that uses the value of `app.update.channel` pref unless `toolkit.telemetry.overrideUpdateChannel` is set:

https://searchfox.org/mozilla-central/rev/37663bb87004167184de6f2afa6b05875eb0528e/toolkit/components/telemetry/app/TelemetryUtils.jsm#393-399

Is there any value in doing this here, or is the "override channel" pref just for test purposes?

Reply:
It's not only for tests, at least [the ASAN Nightly build uses that](https://bugzilla.mozilla.org/show_bug.cgi?id=1471647).

Reply:
Hm, I think we probably don't need to ship this to the ASAN build :) So I ended up using `UpdateUtils` to get the channel just as Telemetry does... I am a little nervous about calling Telemetry code from this add-on (is it possible that things won't work if it is totally disabled?), but if you have strong feelings I can either do that or honor the override pref etc.